File: src/main/java/com/vah/let/sort/QuicklySort.java
package com.vah.let.sort;

import com.vah.let.algorithm.dynamicplanning.Rob;

import java.util.Arrays;

/**
 * @Description
 * @Author vahuang
 **/
public class QuicklySort {
    public static void main(String[] args) {
//        key = 5
        int[] arr = {5, 2, 3, 1, 4, 6, 7, 9, 8};
//        left = 0; right = 4
//        int[] arr = {4, 2, 3, 1, 4, 6, 7, 9, 8};
//        left = 4 right = 4
//        int[] arr = {4, 2, 3, 1, 5, 6, 7, 9, 8};

        quicklySort(arr, 0, arr.length - 1);
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
    }
    public static void quicklySort(int[] arr, int left, int right) {
        if (left < right) {
            int aStandardPoint = qquicklySort(arr, left, right);
            quicklySort(arr, 0,left - 1);
            quicklySort(arr, left + 1, right);
        }
    }

    private static int qquicklySort(int[] arr, int left, int right) {
        // 选取一个key  作为标准点
        int key = arr[left];
        while (left < right) {
            // 高位向低位移动 直到这个数比基数小
            while (left < right && arr[right] >= key) {
                right--;
            }
            arr[left] = arr[right];
            System.out.println(String.format("left:%s;right:%s", left, right));
            pr(arr);
            // 低位向高位移动，直到这个数比基数大
            while (left < right && arr[left] <= key) {
                left++;
            }
            arr[right] = arr[left];
            System.out.println(String.format("left:%s;right:%s", left, right));
            pr(arr);
            arr[left] = key;
            System.out.println(String.format("left:%s;right:%s", left, right));
            pr(arr);
        }
        return left;
    }

    private static void pr(int [] arr) {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }


}

File: src/main/kotlin/com/vah/let/algorithm/sort/QuicklySort.kt
package com.vah.let.algorithm.sort;

/**
 *@Description 快速排序
 *@Author HuangJiang
 **/
class QuicklySort {
}

fun main() {
    val arr = intArrayOf(8, 9, 1, 7, 2, 3, 5, 4, 6, 0)
    quicklySort(arr, 0, arr.size - 1)
    arr.forEach { print("$it,") }
}

fun quicklySort(array: IntArray, low: Int, high: Int) {
    if (low < high) {
        val p = getPartion(array, low, high)
        quicklySort(array, low, p - 1)
        quicklySort(array, p + 1, high)
    }
}

fun getPartion(array: IntArray, low: Int, high: Int): Int {
    val key = array[low]
    var high = high
    var low = low
    while (low < high) {
        while (low < high && array[high] >= key) {
            high--
        }
        array[low] = array[high]
        while (low < high && array[low] < key) {
            low++
        }
        array[high] = array[low]
    }
    array[low] = key
    return high
}
